wal-vdom/src/virtual_dom/vnode.rs

```diff
@@ -1,24 +1,172 @@
-use gloo::console::__macro::JsValue;
+use gloo::{console::log, utils::document};
 use serde::Serialize;
-use web_sys::Node;
+use web_sys::{Element, Node, Text};
 
 use super::{VElement, VText};
 
-#[derive(Serialize)]
+#[derive(Debug, Serialize)]
 pub enum VNode {
-    Element{
-        velement: VElement,
+    Element {
+        #[serde(skip_serializing)]
+        concrete: Option<Element>,
+        virt: VElement,
     },
-    Text{
-        vtext: VText,
+    Text {
+        #[serde(skip_serializing)]
+        concrete: Option<Text>,
+        virt: VText,
     },
 }
 
 impl VNode {
-    pub fn render(&self) -> Result<Node, JsValue> {
+    pub fn patch(&mut self, last: Option<VNode>, ancestor: &Node) {
         match self {
-            VNode::Element { velement } => velement.render(),
-            VNode::Text { vtext } => vtext.render(),
+            VNode::Element {
+                ref mut concrete,
+                virt,
+            } => VNode::patch_element(concrete, virt, last, ancestor),
+            VNode::Text {
+                ref mut concrete,
+                virt,
+            } => VNode::patch_text(concrete, virt, last, ancestor),
+        };
+    }
+
+    fn patch_element(
+        concrete: &mut Option<Element>,
+        virt: &mut VElement,
+        last: Option<VNode>,
+        ancestor: &Node,
+    ) {
+        log!("Patching element");
+        let mut old_virt: Option<VElement> = None;
+        match last {
+            // First creating the node
+            Some(VNode::Element { concrete: None, .. })
+            | Some(VNode::Text { concrete: None, .. })
+            | None => {
+                log!("\tCreating the node from the bottom");
```

Comment:
to sie wydaje dziwne. Jesli przesledzimy ten kod to wychodzi na to ze tworzymy element za pomocą create_element, ale nadal old_virt jest None. Sprawia to że w linii 80 w funkcji render znowu będzimy tworzyć element. Imo to jeszcze można się zastanowić nad tym całym mechanizmem